File: campus-navigation/campus-navigation/audit_data.py

import json
import re

# File path
file_path = r'e:\Project\full_project_backup\src\campus-navigation\src\data\campusData.js'

def audit_campus_data():
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Very rough extraction of JS objects using regex (since it's a JS file)
    # This is brittle, but helps find obvious missing IDs or broken refs.
    
    # 1. Search for ROOM IDs in FLOOR_DATA
    room_ids_in_floors = re.findall(r"id:\s*['\"]([^'\"]+)['\"]", content)
    
    # 2. Search for spatial IDs
    spatial_ids = re.findall(r"spatialId:\s*['\"]([^'\"]+)['\"]", content)
    
    # 3. Search for nodes in ROUTING_NODES
    nodes = re.findall(r"ROUTING_NODES = \{([\s\S]+?)\};", content)
    node_ids = []
    if nodes:
        node_ids = re.findall(r"['\"]?([^'\":\s]+)['\"]?:\s*\{", nodes[0])

    # 4. Search for edges in ROUTING_EDGES
    edges = re.findall(r"ROUTING_EDGES = \[([\s\S]+?)\];", content)
    broken_edges = []
    if edges:
        edge_data = re.findall(r"from:\s*['\"]([^'\"]+)['\"],\s*to:\s*['\"]([^'\"]+)['\"]", edges[0])
        for fr, to in edge_data:
            if fr not in node_ids:
                broken_edges.append(f"Broken edge 'from': {fr}")
            if to not in node_ids:
                broken_edges.append(f"Broken edge 'to': {to}")

    # 5. Check ALL_DEPARTMENTS
    depts = re.findall(r"ALL_DEPARTMENTS = \[([\s\S]+?)\];", content)
    broken_depts = []
    if depts:
        # Check if dept node (id) matches a room id or routing node
        dept_ids = re.findall(r"id:\s*['\"]([^'\"]+)['\"]", depts[0])
        for d_id in dept_ids:
            # Note: ALL_DEPARTMENTS 'id' is for search, but 'spatialId' connects it.
            pass

    print(f"Audit Results:")
    print(f"Total Room IDs: {len(room_ids_in_floors)}")
    print(f"Total Spatial IDs: {len(spatial_ids)}")
    print(f"Total Routing Nodes: {len(node_ids)}")
    print(f"Broken Edges: {len(broken_edges)}")
    for be in broken_edges: print(f" - {be}")
    
    # Check for duplicate spatial IDs
    dupes = set([x for x in spatial_ids if spatial_ids.count(x) > 1])
    print(f"Duplicate Spatial IDs: {len(dupes)}")
    for d in dupes: print(f" - {d}")

audit_campus_data()

File: campus-navigation/campus-navigation/audit-data.js

import { ROUTING_NODES, ROUTING_EDGES, ALL_DEPARTMENTS, FLOOR_DATA, BUILDINGS } from './src/data/campusData.js';

function audit() {
  console.log('--- Campus Navigation Data Audit ---');
  let errors = 0;

  // 1. Check for Duplicate Node IDs
  const nodes = Object.keys(ROUTING_NODES);
  console.log(`Total Routing Nodes: ${nodes.length}`);

  // 2. Check for Broken Edges
  console.log('Checking Edges...');
  ROUTING_EDGES.forEach((edge, i) => {
    if (!ROUTING_NODES[edge.from]) {
      console.error(`Error: Edge ${i} has broken 'from' node: ${edge.from}`);
      errors++;
    }
    if (!ROUTING_NODES[edge.to]) {
      console.error(`Error: Edge ${i} has broken 'to' node: ${edge.to}`);
      errors++;
    }
  });

  // 3. Check for Searchable Departments vs Nodes
  console.log('Checking Departments...');
  ALL_DEPARTMENTS.forEach(dept => {
    let foundInFloors = false;
    Object.values(FLOOR_DATA).forEach(building => {
      Object.values(building).forEach(floor => {
        if (floor.rooms) {
          if (floor.rooms.find(r => r.spatialId === dept.spatialId)) foundInFloors = true;
        }
      });
    });
    if (!foundInFloors) {
      console.warn(`Warning: Department ${dept.name} (${dept.spatialId}) not found in any Floor Plan.`);
      errors++;
    }
  });

  // 4. Check for duplicate spatial IDs in Floor Plans
  const allSpatialIds = [];
  Object.entries(FLOOR_DATA).forEach(([bId, floors]) => {
    Object.entries(floors).forEach(([fId, data]) => {
      if (data.rooms) {
        data.rooms.forEach(room => {
          if (room.spatialId) {
            if (allSpatialIds.includes(room.spatialId)) {
              console.error(`Error: Duplicate Spatial ID found: ${room.spatialId} (Bldg ${bId}, Floor ${fId})`);
              errors++;
            }
            allSpatialIds.push(room.spatialId);
          }
        });
      }
    });
  });

  // 5. Check Building vs Floor levels
  console.log('Checking Building/Floor consistency...');
  Object.entries(BUILDINGS).forEach(([bId, bData]) => {
    if (!bData.levels) {
      console.error(`Error: Building ${bId} has no 'levels' defined.`);
      errors++;
    } else {
      bData.levels.forEach(lvl => {
        if (!FLOOR_DATA[bId] || !FLOOR_DATA[bId][lvl]) {
          console.warn(`Warning: Building ${bId} Level ${lvl} is defined in BUILDINGS but missing from FLOOR_DATA.`);
          errors++;
        }
      });
    }
  });

  // 6. Check Department Building references
  console.log('Checking Department Building references...');
  ALL_DEPARTMENTS.forEach(dept => {
    if (!BUILDINGS[dept.building]) {
      console.error(`Error: Department ${dept.name} references non-existent building: ${dept.building}`);
      errors++;
    }
  });

  console.log(`\nAudit Complete. Total Potential Bugs Found: ${errors}`);
}

audit();
